shardmanager.go

```diff
@@ -84,19 +85,56 @@ func NewShardManager(
 
 // General
 
-func (n *ShardManager) connectShardNodes(ctx context.Context, shardID ShardIDType) error {
+// connectShardNodes connects to new shard peers until
+// we have at least `numShardPeerToConnect` number of shard peers.
+// Note that if we don't have the shard peer's listening address,
+// a DHT query will be made which will change the network topology.
+func (n *ShardManager) connectShardNodes(ctx context.Context, numShardPeerToConnect int, shardID ShardIDType) error {
 	// Add span for connectShardNodes of ShardManager
 	spanctx := logger.Start(ctx, "ShardManager.connectShardNodes")
 	defer logger.Finish(spanctx)
 	logger.SetTag(spanctx, "shard", shardID)
 
-	pinfos, err := n.discovery.FindPeers(spanctx, strconv.FormatInt(shardID, 10))
+	//Get shard peers that we already connected to
+	connectedPIDs := n.pubsubService.ListPeers(getCollationsTopic(shardID))
+	// Quit if we already have enough shard peers
+	if len(connectedPIDs) >= numShardPeerToConnect {
+		return nil
+	}
+	// Find peers that also subscribed to the shard
+	foundPeers, err := n.discovery.FindPeers(spanctx, strconv.FormatInt(shardID, 10))
 	if err != nil {
 		logger.SetErr(spanctx, fmt.Errorf("Failed to find peers in shard %v", shardID))
 		logger.Errorf("Failed to find peers in shard %v", shardID)
 		return err
 	}
 
+	pinfos := make([]pstore.PeerInfo, 0)
+	// First filter out peers that we already connected to
+	for _, p := range foundPeers {
+		found := false
+		for _, pid := range connectedPIDs {
```

Comment:
Maybe it will be more efficient to make a `map` with `connectedPIDs`, and look up it when iterate through `foundPeers`?